### ISSUE TITLE
WaterfallDialog throws InvalidCastException when using CosmosDbStorage 

### ISSUE THREAD

**Issue by lauren-mills** [Microsoft/botbuilder-dotnet#871]
When I'm using CosmosDbStorage in the BotState middleware I see this exception at the line listed:

Error: InvalidCastException: Unable to cast object of type 'System.Int64' to type 'System.Int32'.

https://github.com/Microsoft/botbuilder-dotnet/blob/5764c52b03c31ca2a70c0730046cd17ef50d8c5d/libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs#L72

I have no issues when using MemoryStorage.

**Stack Trace:**

```plain
System.InvalidCastException: Unable to cast object of type 'System.Int64' to type 'System.Int32'.
   at Microsoft.Bot.Builder.Dialogs.WaterfallDialog.<DialogResumeAsync>d__7.MoveNext() in D:\a\1\s\libraries\Microsoft.Bot.Builder.Dialogs\WaterfallDialog.cs:line 65
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.DialogContext.<EndAsync>d__18.MoveNext() in D:\a\1\s\libraries\Microsoft.Bot.Builder.Dialogs\DialogContext.cs:line 173
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Prompt`1.<DialogContinueAsync>d__5.MoveNext() in D:\a\1\s\libraries\Microsoft.Bot.Builder.Dialogs\Prompts\Prompt.cs:line 101
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.DialogContext.<ContinueAsync>d__17.MoveNext() in D:\a\1\s\libraries\Microsoft.Bot.Builder.Dialogs\DialogContext.cs:line 130
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at EnterpriseBot.MainDialog.<OnTurnAsync>d__12.MoveNext() in C:\Users\lamil\source\repos\CAISolutions\EnterpriseBot\EnterpriseBot\MainDialog\MainDialog.cs:line 66
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at EnterpriseBot.DispatcherMiddleware.<OnTurnAsync>d__6.MoveNext() in C:\Users\lamil\source\repos\CAISolutions\EnterpriseBot\EnterpriseBot\Middleware\DispatcherMiddleware.cs:line 137
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at EnterpriseBot.ContentModeratorMiddleware.<OnTurnAsync>d__4.MoveNext() in C:\Users\lamil\source\repos\CAISolutions\EnterpriseBot\EnterpriseBot\Middleware\ContentModeratorMiddleware.cs:line 76
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.TranscriptLoggerMiddleware.<OnTurnAsync>d__4.MoveNext() in D:\a\1\s\libraries\Microsoft.Bot.Builder\TranscriptLoggerMiddleware.cs:line 105
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.BotStateSet.<OnTurnAsync>d__5.MoveNext() in D:\a\1\s\libraries\Microsoft.Bot.Builder\BotStateSet.cs:line 52
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.MiddlewareSet.<ReceiveActivityWithStatusAsync>d__4.MoveNext() in D:\a\1\s\libraries\Microsoft.Bot.Builder\MiddlewareSet.cs:line 67
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.BotAdapter.<RunPipelineAsync>d__13.MoveNext() in D:\a\1\s\libraries\Microsoft.Bot.Builder\BotAdapter.cs:line 167
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Bot.Builder.BotAdapter.<RunPipelineAsync>d__13.MoveNext() in D:\a\1\s\libraries\Microsoft.Bot.Builder\BotAdapter.cs:line 160
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.BotFrameworkAdapter.<ProcessActivityAsync>d__12.MoveNext() in D:\a\1\s\libraries\Microsoft.Bot.Builder\BotFrameworkAdapter.cs:line 207
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.BotFrameworkAdapter.<ProcessActivityAsync>d__11.MoveNext() in D:\a\1\s\libraries\Microsoft.Bot.Builder\BotFrameworkAdapter.cs:line 180
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Integration.AspNet.Core.Handlers.BotMessageHandler.<ProcessMessageRequestAsync>d__0.MoveNext() in D:\a\1\s\libraries\integration\Microsoft.Bot.Builder.Integration.AspNet.Core\BotMessageHandler.cs:line 26
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Integration.AspNet.Core.Handlers.BotMessageHandlerBase.<HandleAsync>d__2.MoveNext() in D:\a\1\s\libraries\integration\Microsoft.Bot.Builder.Integration.AspNet.Core\BotMessageHandlerBase.cs:line 71
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Builder.Extensions.MapMiddleware.<Invoke>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.<Invoke>d__7.MoveNext()
```

**Comment by drub0y** [comment]
@lauren-mills hey, thanks for reporting this! Can I ask a favor? Can you update your original text with the full stacktrace of where the exception is thrown? You can grab this by just calling `ToString()` on the exception when it happens. That'll likely help us pinpoint the issue without having to ask you for a bunch of other details so we can figure out a repro.

**Comment by drub0y** [comment]
Thanks for posting that stack trace @lauren-mills!

So what is happening here is that, because a `Dictionary<string, object>` is being used for dialog state, when the JSON is being deserialized and the serializer encounters an `number` value it only knows that the value needs to "fit" into an `object` which obviously doesn't tell it anything about how wide the integer can/should be. Therefore I believe it will just make the assumption that the value could be the widest possible integer type which is the `Int64` since the `number` type in JavaScript is treated as a double-precision 64-bit value and _not_ technically an integer, and that's how we end up here. It could _also_ be that the deserializer looks at the "bitness" of the process and goes with the widest possible integer type for that bitness, but, either way, the assumption that it will come back out as an `Int32` is wrong.

**Comment by darrenj** [comment]
Thanks - just ran into this myself and did the same fix to get me moving

**Comment by EricDahlvang** [comment]
A good explanation of this occurs can be found here: https://ericlippert.com/2009/03/03/representation-and-identity/

**Comment by stevengum** [comment]
A fix was merged into master a short while ago via #891, closing issue.
